contest_service/src/main.rs

```diff
@@ -1,207 +1,273 @@
+use futures::stream::StreamExt;
+use std::convert::TryFrom;
+
+use mappings::chat::Message;
 use mongodb::{
     bson::{doc, Document},
-    options::ClientOptions,
+    options::{ClientOptions, UpdateOptions},
     Client,
 };
 use protos::service::contest::{contest_server::*, *};
 use protos::utils::*;
 use tonic::{transport::*, Request, Response, Status};
 
+mod mappings;
+
 #[cfg(test)]
 mod tests;
 
+// TODO: remove credentials to connect to db.
 const CONNECTION_STRING: &str = "mongodb://root:example@contest_service_db:27017/";
 
+fn internal_error<T>(e: T) -> Status
+where
+    T: std::fmt::Debug,
+{
+    Status::internal(format!("{:?}", e))
+}
+
 #[derive(Debug)]
 pub struct ContestService {
-    client: Client,
+    db_client: Client,
 }
 
 impl ContestService {
     async fn new() -> Result<Self, Box<dyn std::error::Error>> {
         Ok(Self {
-            client: Client::with_options(ClientOptions::parse(CONNECTION_STRING).await?)?,
+            db_client: Client::with_options(ClientOptions::parse(CONNECTION_STRING).await?)?,
         })
     }
-    fn get_contests_collection(&self) -> mongodb::Collection<Document> {
-        let db = self.client.database("contestdb");
-        db.collection::<Document>("contests")
+
+    /// Do not call this function, call get_*_collection or get_contest_metadata instead
+    fn get_collection(&self, collection_name: &str) -> mongodb::Collection<Document> {
+        let db = self.db_client.database("contestdb");
+        db.collection::<Document>(collection_name)
```

Comment:
Consider populating the collection with the correct type from the beginning instead of the general Document.
As of now, we implement the `From` trait to convert between the `mongodb::bson::Document` and the rust object.
Take a look at https://docs.rs/mongodb/2.0.0-beta/mongodb/struct.Database.html#method.collection

Reply:
Ah damn, seems almost like I haven't thoroughly read the doc

Reply:
I've implemented, but not in:
- `User`: because I hash the password on serialization, although I'm quite sure this can be accomplished using https://serde.rs/field-attrs.html
- `Problem`, because additional work was needed and was not worth the effort imho